dll/win32/shdocvw/mrulist.cpp

```diff
@@ -397,6 +404,145 @@ HRESULT CMruBase::_DeleteValue(LPCWSTR pszValue)
     return SHDeleteValueW(m_hKey, NULL, pszValue);
 }
 
+HRESULT CMruBase::_UseEmptySlot(UINT *piSlot)
+{
+    if (!m_bChecked)
+        _CheckUsedSlots();
+
+    if (!m_cSlotRooms)
+        return E_FAIL;
+
+    UINT iSlot = 0;
+    for (SLOTITEMDATA *pItem = m_pSlots; (pItem->dwFlags & SLOT_SET); ++pItem)
+    {
+        if (++iSlot >= m_cSlotRooms)
+            return E_FAIL;
+    }
+
+    m_pSlots[iSlot].dwFlags |= SLOT_SET;
+    *piSlot = iSlot;
+    ++m_cSlots;
+
+    return S_OK;
+}
+
+class CMruShortList
+    : public CMruBase
+{
+protected:
+    LPWSTR m_pszSlotData = NULL;
+
+    HRESULT _InitSlots() override;
+    void _SaveSlots() override;
+    UINT _UpdateSlots(UINT iSlot) override;
+    void _SlotString(DWORD dwSlot, LPWSTR psz, DWORD cch) override;
+    HRESULT _GetSlot(UINT iSlot, UINT *puSlot) override;
+    HRESULT _RemoveSlot(UINT iSlot, UINT *puSlot) override;
+    friend class CMruLongList;
+
+public:
+    CMruShortList()
+    {
+    }
+
+    ~CMruShortList() override
+    {
+        m_pszSlotData = (LPWSTR)::LocalFree(m_pszSlotData);
+    }
+};
+
+HRESULT CMruShortList::_InitSlots()
+{
+    DWORD cbData = (m_cSlotRooms + 1) * sizeof(WCHAR);
+    m_pszSlotData = (LPWSTR)LocalAlloc(LPTR, cbData);
+    if (!m_pszSlotData)
+        return E_OUTOFMEMORY;
+
+    if (SHGetValueW(m_hKey, NULL, L"MRUList", NULL, m_pszSlotData, &cbData) == ERROR_SUCCESS)
+        m_cSlots = (cbData / sizeof(WCHAR)) - 1;
+
+    m_pszSlotData[m_cSlots] = 0;
+    return S_OK;
+}
+
+void CMruShortList::_SaveSlots()
+{
+    if (m_bNeedSave)
+    {
+        DWORD cbData = (m_cSlots + 1) * sizeof(WCHAR);
+        SHSetValueW(m_hKey, NULL, L"MRUList", REG_SZ, m_pszSlotData, cbData);
+        m_bNeedSave = FALSE;
+    }
+}
+
+UINT CMruShortList::_UpdateSlots(UINT iSlot)
+{
+    UINT iData, cDataToMove = iSlot;
+
+    if (iSlot == m_cSlots)
+    {
+        if (SUCCEEDED(_UseEmptySlot(&iData)))
+        {
+            ++cDataToMove;
+        }
+        else
+        {
+            iData = m_pszSlotData[m_cSlots - 1] - L'a';
```

Comment:
What is happening there? Missing comment here or in similar code that manipulate this m_pszSlotData: Explanation of the format of the stored data.

Reply:
MRUList uses lowercase alphabet for history of most recently used items.

Reply:
This may be of interest to mention that in a comment in the code.

Reply:
Added some comments.